### ISSUE TITLE
Optionally preserve current stack

### ISSUE THREAD

**Issue by fregante** [sindresorhus/serialize-error#98]
An issue I found when throwing deserialized errors is that the thrown error has no reference to this `throw` point.

I think the "correct" way to throw serialized errors would be something like:

```js
throw new Error(serialized.message, {
	cause: deserializeError(serialized),
});
```

Two drawbacks:

- `new Error` doesn't benefit from https://github.com/sindresorhus/serialize-error/pull/70
- this isn't necessarily what people want or expect by default

So how about:

```js
throw deserializeError(serialized, {preserveStack: true});
```

**Comment by sindresorhus** [comment]
I'm ok with adding this, but I think the option is ambiguous. I'm asking myself, which stack is it preserving.

**Comment by fregante** [comment]
I think it's going to be hard to find encapsulate the entire behavior and reason d'être in a property name, especially since the main verb here is "deserialize error" rather than "create error"

Names:

- `asCause` (doesn't explain why)
- `captureCurrentStack` (doesn't say a new error is created)
- `newError` (technically `deserialize` is already creating a `new Error()`)
- `wrapped` (good description of the result but says nothing about the behavior)

I think `asCause` gives the clearest picture of the result, assuming you're aware of `Error#cause`, even if it doesn't explain the point of doing so.

**Comment by sindresorhus** [comment]
> I think asCause gives the clearest picture of the result, assuming you're aware of Error#cause, even if it doesn't explain the point of doing so.

I assumed you would just want the `stack` property replaced on the deserialized error with the current stack. It now sounds like you want the option to wrap it in a new error and add the deserialized one as `cause`?

**Comment by fregante** [comment]
Correct, check my `new Error` example. I want exactly that, but safer, with less code and with the ability to use the right error constructor automatically.

**Comment by sindresorhus** [comment]
How about calling the option `wrap`?

**Comment by fregante** [comment]
I think "deserialize error as cause" makes more sense (`{asCause: true}`). Or "deserialize error and capture stack", to use node terminology analogous to Error.captureStack(), which takes it from the current line (`{captureStack: true}`)

"Wrap" doesn't really say anything to me

**Comment by fregante** [comment]
I'm ok without this option. Creating a new error with the same error type and error message seems redundant anyway. 

Anyone who wants to preserve the current stack should throw an error with an error relevant to the current throw place, for example "MessagingError: Received an error from the recipient. Cause: `deserializeError(error)`"
